schneiderWallbox/integrationpluginschneiderwallbox.cpp

```diff
@@ -0,0 +1,546 @@
+        thing->setStateValue("currentPower", stationPowerTotalWatt);
+        thing->setStateValue("charging", stationPowerTotalWatt > 20);
+    });
+
+    connect(schneiderWallboxTcpConnection, &SchneiderWallboxModbusTcpConnection::errorStatusChanged, this, [thing, this](quint32 errorBits){
+        qCDebug(dcSchneiderElectric()) << thing << "Error state changed" << errorBits ;
+        setErrorMessage(thing, errorBits);
+    });
+
+    schneiderWallboxTcpConnection->connectDevice();
+
+    connect(schneiderWallbox, &SchneiderWallbox::phaseCountChanged, this, [thing, this](quint16 phaseCount){
+        qCDebug(dcSchneiderElectric()) << thing << "Phase count changed" << phaseCount ;
+        setPhaseCount(thing, phaseCount);
+    });
+
+    /* Probably not needed.
+    connect(schneiderWallbox, &SchneiderWallbox::currentPowerChanged, this, [thing, this](double currentPower){
+        qCDebug(dcSchneiderElectric()) << thing << "Current power changed" << currentPower ;
+        setCurrentPower(thing, currentPower);
+        thing->setStateValue("currentPower", currentPower);
+        thing->setStateValue("charging", currentPower > 20);
+    });
+    */
+}
+
+
+void IntegrationPluginSchneiderWallbox::postSetupThing(Thing *thing)
+{
+    qCDebug(dcSchneiderElectric()) << "Post setup" << thing->name();
+    if (thing->thingClassId() != schneiderEVlinkThingClassId) {
+        qCWarning(dcSchneiderElectric()) << "Thing class id not supported" << thing->thingClassId();
+        return;
+    }
+
+    if (!m_pluginTimer) {
+        qCDebug(dcSchneiderElectric()) << "Starting plugin timer...";
+        m_pluginTimer = hardwareManager()->pluginTimerManager()->registerTimer(3);
+        connect(m_pluginTimer, &PluginTimer::timeout, this, [this] {
+            foreach(Thing *thing, myThings()) {
+                if (m_monitors.value(thing)->reachable()) {
+                    qCDebug(dcSchneiderElectric()) << "Updating" << thing->name();
+                    m_schneiderDevices.value(thing)->modbusTcpConnection()->update();
+                    m_schneiderDevices.value(thing)->update();
+                } else {
+                    qCDebug(dcSchneiderElectric()) << thing->name() << "isn't reachable. Not updating.";
+                }
+            }
+        });
+
+        //m_pluginTimer->start();  apparently not needed anymore.
+    }
+}
+
+void IntegrationPluginSchneiderWallbox::thingRemoved(Thing *thing)
+{
+    qCDebug(dcSchneiderElectric()) << "Removing device" << thing->name();
+    if (m_schneiderDevices.contains(thing)) {
+        disconnect(m_schneiderDevices.value(thing)->modbusTcpConnection(), NULL, thing, NULL);  // This is needed to prevent crash on thing remove.
+        // m_schneiderDevices.value(thing)->modbusTcpConnection()->disconnect(thing);  // This should do exactly the same thing as the line above. Somehow it does not. This line does not prevent the crash.
```

Comment:
yes. `disconnect(QObject*)` is a different thing than `disconnect(QObject *, null, QObject *, null)`. As per Qt documentation, the former will disconnect all signals from the given object (not disconnecting any slots) while the latter will disconnect all signals of the first object from slots of the second object.

In any case. IMO you should be able to drop all of this by changing the line below from

```
m_schneiderDevices.take(thing)->deleteLater()
```
to
```
delete m_schneiderDevice.take(thing):
```
That said, I did not verify if that may change stuff in some other place but it should be ok.